crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,552 @@
+    let extra = properties.extra(&all_allowed);
+    if !extra.is_empty() {
+        return Err(DeviceError::UnexpectedDeviceProperties(
+            extra.iter().sorted().join(", "),
+        ));
+    }
+    Ok(())
+}
+
+// Common sensor infrastructure
+
+/// Basic information about the sensor.
+///
+/// It is typically used as a field in structs implementing sensor devices.
+#[derive(Debug)]
+pub struct Sensor {
+    /// The sensor name (possibly truncated) as reported to the guest.
+    pub name: String,
+    /// Whether the sensor is enabled.
+    ///
+    /// Sensors can be enabled and disabled using SCMI.  [Sensor]s created
+    /// using [Sensor::new] are disabled initially.
+    enabled: bool,
+}
+
+impl Sensor {
+    pub fn new(properties: &DeviceProperties, default_name: &str) -> Self {
+        let name = properties.get("name").unwrap_or(default_name);
+        Self {
+            name: name.to_owned(),
+            enabled: false,
+        }
+    }
+}
+
+/// Common base that sensor devices can use to simplify their implementation.
+///
+/// To add a new kind of sensor bindings, you must implement
+/// [crate::scmi::ScmiDevice], define [DeviceSpecification] and add it to
+/// [NameDeviceMapping] created in [available_devices].  You can do it fully
+/// yourself or use this trait to simplify the implementation.
+///
+/// The trait is typically used as follows:
+///
+/// ```rust
+/// struct MySensor {
+///     sensor: Sensor,
+///     // other fields as needed
+/// }
+///
+/// impl SensorT for MySensor {
+///     // provide trait functions implementation as needed
+/// }
+///
+/// impl MySensor {
+///     pub fn new_device(properties: &DeviceProperties) -> MaybeDevice {
+///         check_device_properties(properties, &[], &["name"])?;
+///         let sensor = Sensor::new(properties, "mydevice");
+///         let my_sensor = MySensor { sensor };
+///         let sensor_device = SensorDevice(Box::new(my_sensor));
+///         Ok(Box::new(sensor_device))
+///     }
+/// }
+/// ```
+///
+/// See [crate::devices::fake::FakeSensor] implementation for an example.
+pub trait SensorT: Send {
+    /// Returns the inner [Sensor] instance, immutable.
+    fn sensor(&self) -> &Sensor;
+    /// Returns the inner [Sensor] instance, mutable.
+    fn sensor_mut(&mut self) -> &mut Sensor;
+
+    /// Performs any non-default initialization on the sensor.
+    ///
+    /// If the initialization fails, a corresponding error message is
+    /// returned.
+    fn initialize(&mut self) -> Result<(), DeviceError> {
+        Ok(())
+    }
+
+    /// Returns the id of the SCMI protocol used to communicate with the
+    /// sensor.
+    ///
+    /// Usually no need to redefine this.
+    fn protocol(&self) -> ProtocolId {
+        SENSOR_PROTOCOL_ID
+    }
+
+    /// Returns an error message about invalid property `name`.
+    ///
+    /// Usually no need to redefine this.
+    fn invalid_property(&self, name: &String) -> Result<(), String> {
```

Comment:
```suggestion
    fn invalid_property(&self, name: &str) -> Result<(), String> {
```
and maybe using an error type here instead of a string?

Reply:
Done.